engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java

```diff
@@ -136,201 +98,166 @@ public ReplayStateMachine(final ProcessingContext context) {
-          updateStateUntilDone();
-        });
   }
 
-  private void replayEvent(final TypedRecord<?> currentEvent) {
-    // skip events if the state changes are already applied to the state in the snapshot
-    if (currentEvent.getSourceRecordPosition() > snapshotPosition) {
-      eventApplier.applyState(
-          currentEvent.getKey(), currentEvent.getIntent(), currentEvent.getValue());
+  /**
+   * Tries to apply the current read event to the state. If this method is called multiple times,
+   * because of previous errors, then it will reset the current transaction before it tries to apply
+   * the event again.
+   *
+   * @return true on success
+   * @throws Exception if the applying of the event fails
+   */
+  private boolean tryToApplyCurrentEvent() throws Exception {
+    final boolean onRetry = zeebeDbTransaction != null;
+    if (onRetry) {
+      zeebeDbTransaction.rollback();
     }
-  }
 
-  private void updateStateUntilDone() {
-    final ActorFuture<Boolean> retryFuture =
-        updateStateRetryStrategy.runWithRetry(
-            () -> {
-              zeebeDbTransaction.commit();
-              zeebeDbTransaction = null;
-              return true;
-            },
-            abortCondition);
-
-    actor.runOnCompletion(
-        retryFuture,
-        (bool, throwable) -> {
-          // update state should be retried endless until it worked
-          assert throwable == null : "On replay there shouldn't be any exception thrown.";
-          onRecordReplayed(currentEvent);
+    zeebeDbTransaction = transactionContext.getCurrentTransaction();
+    zeebeDbTransaction.run(
+        () -> {
+          if (typedEvent.getSourceRecordPosition() > snapshotPosition) {
+            eventApplier.applyState(
+                typedEvent.getKey(), typedEvent.getIntent(), typedEvent.getValue());
+          }
+          lastProcessedPositionState.markAsProcessed(currentEvent.getPosition());
         });
-  }
-
-  private void onRecordReplayed(final LoggedEvent currentEvent) {
-    if (currentEvent.getPosition() >= lastFollowUpEventPosition) {
-      LOG.info(LOG_STMT_REPLAY_FINISHED, currentEvent.getPosition());
 
-      // reset the position to the first event where the processing should start
-      logStreamReader.seekToNextEvent(lastSourceEventPosition);
-
-      onRecovered(lastSourceEventPosition);
-    } else {
-      actor.submit(this::replayNextEvent);
-    }
+    return true;
   }
 
-  private void onRecovered(final long lastProcessedPosition) {
-    keyGeneratorControls.setKeyIfHigher(highestRecordKey);
-
-    recoveryFuture.complete(lastProcessedPosition);
+  /**
+   * Commits the current transaction. The transaction holds the state changes, which have been
+   * caused by applying the current event.
+   *
+   * @return true on success
+   * @throws Exception if the commit fails
+   */
+  private boolean commitCurrentAppliedStateChanges() throws Exception {
```

Comment:
(optional) We could rename the method to `tryToCommitStateChanges()`.

This name would align with `tryToApplyCurrentEvent()`.